src/clr-angular/utils/focus-trap/focus-trap.directive.ts

```diff
@@ -39,13 +46,42 @@ export class FocusTrapDirective implements AfterViewInit, OnDestroy {
     this.renderer.setAttribute(this.el.nativeElement, 'tabindex', '0');
   }
 
+  @Input('clrFocusTrap')
+  config: FocusTrapConfig = {
+    local: false,
+  };
+
   @HostListener('document:focusin', ['$event'])
   onFocusIn(event: any) {
+    if (this.focusTrapsTracker.current !== this || !isPlatformBrowser(this.platformId) || this.localFocusEscaped) {
+      return;
+    }
     const nativeElement: HTMLElement = this.el.nativeElement;
-
-    if (this.focusTrapsTracker.current === this && event.target && !nativeElement.contains(event.target)) {
+    if (!this.config.local && event.target && !nativeElement.contains(event.target)) {
+      // When the focus trap is global, always steal focus back if it goes outside
+      nativeElement.focus();
+    } else if (event.target === this.bottomReboundEl && nativeElement.contains(this.previousTrappedActiveElement)) {
+      // When the focus trap is local, if the user navigates via keyboard to the end element from within the trap, move to top
       nativeElement.focus();
+    } else if (event.target === this.topReboundEl) {
+      // When the focus trap is local, if the user navigates via keyboard back to start element from within the trap, move to bottom
+      const focusableElements: NodeList = nativeElement.querySelectorAll(
+        'button, [href], input, select, textarea, [tabindex]:not([tabindex="-1"])'
```

Comment:
This was a concern of mine, and I don't have a better solution. 

1. This is true, but it is also possible to create a selector here that catches as much as it can, and if there is a quirky behavior folks can workaround by adding a `tabindex`. Seems like an unlikely scenario for 99% of use.
2. But why don't you focus on a disabled element? You can in forms to know about it, this isn't a problem of mine.
3. Hidden elements is a concern, yet we could advise to remove it from the DOM instead of hiding through CSS.

Reply:
This part is trying to solve SHIFT+TAB issue which is not related to the core issue. This PR can solve the greedy focus trap issue without this part. If we want to solve SHIFT+TAB problem, we should tackle that in a separate PR.

Reply:
Alright, we can drop this reverse rebound flow for now.